File: app/src/main/java/zhan/rnadomgirdanimation/refresh/IDiffCallBack.java
package zhan.rnadomgirdanimation.refresh;

import android.support.v7.util.DiffUtil;
import java.util.List;

/**
 * Created by zhan on 2016/12/11.
 */

public abstract class IDiffCallBack<T> extends DiffUtil.Callback {

  protected List<T> mOldData;
  protected List<T> mNewData;

  public void setData(List<T> oldData, List<T> newData) {
    mOldData = oldData;
    mNewData = newData;
  }

  @Override public int getOldListSize() {
    return mOldData == null ? 0 : mOldData.size();
  }

  @Override public int getNewListSize() {
    return mNewData == null ? 0 : mNewData.size();
  }

  public abstract boolean isItemsTheSame(int oldItemPosition, int newItemPosition);

  @Override public boolean areItemsTheSame(int oldItemPosition, int newItemPosition) {
    return isItemsTheSame(oldItemPosition, newItemPosition);
  }

  public abstract boolean isContentsTheSame(int oldItemPosition, int newItemPosition);

  @Override public boolean areContentsTheSame(int oldItemPosition, int newItemPosition) {
    return isContentsTheSame(oldItemPosition, newItemPosition);
  }
}

File: app/src/main/java/zhan/rnadomgirdanimation/refresh/IDiffCallBack.kt
package zhan.rnadomgirdanimation.refresh

import androidx.recyclerview.widget.DiffUtil

abstract class IDiffCallBack<T> : DiffUtil.Callback() {

    protected var mOldData: List<T>? = null
    protected var mNewData: List<T>? = null

    fun setData(oldData: List<T>, newData: List<T>) {
        mOldData = oldData
        mNewData = newData
    }

    override fun getOldListSize(): Int {
        return if (mOldData == null) 0 else mOldData!!.size
    }

    override fun getNewListSize(): Int {
        return if (mNewData == null) 0 else mNewData!!.size
    }

    abstract fun isItemsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean

    override fun areItemsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
        return isItemsTheSame(oldItemPosition, newItemPosition)
    }

    abstract fun isContentsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean

    override fun areContentsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
        return isContentsTheSame(oldItemPosition, newItemPosition)
    }
}
